File: dataModule/src/main/java/com/mobnetic/coinguardian/model/market/Bitso.java
package com.mobnetic.coinguardian.model.market;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;

import org.json.JSONArray;
import org.json.JSONObject;

import com.mobnetic.coinguardian.model.CheckerInfo;
import com.mobnetic.coinguardian.model.CurrencyPairInfo;
import com.mobnetic.coinguardian.model.Market;
import com.mobnetic.coinguardian.model.Ticker;
import com.mobnetic.coinguardian.model.currency.Currency;
import com.mobnetic.coinguardian.model.currency.VirtualCurrency;
import com.mobnetic.coinguardian.util.ParseUtils;

public class Bitso extends Market {

	private final static String NAME = "Bitso";
	private final static String TTS_NAME = NAME;
	private final static String URL = "https://api.bitso.com/public/info";
	private final static HashMap<String, CharSequence[]> CURRENCY_PAIRS = new LinkedHashMap<>();
	static {
		CURRENCY_PAIRS.put(VirtualCurrency.BTC, new String[]{
				Currency.MXN
			});
		CURRENCY_PAIRS.put(VirtualCurrency.ETH, new String[]{
				VirtualCurrency.BTC,
				Currency.MXN
			});
		CURRENCY_PAIRS.put(VirtualCurrency.XRP, new String[]{
				VirtualCurrency.BTC,
				Currency.MXN
			});
		CURRENCY_PAIRS.put(VirtualCurrency.BCH, new String[]{
				VirtualCurrency.BTC
			});
		CURRENCY_PAIRS.put(VirtualCurrency.LTC, new String[]{
				VirtualCurrency.BTC,
				Currency.MXN
			});
	}
	
	public Bitso() {
		super(NAME, TTS_NAME, CURRENCY_PAIRS);
	}

	@Override
	public String getUrl(int requestId, CheckerInfo checkerInfo) {
		return URL;
	}
	
	@Override
	protected void parseTickerFromJsonObject(int requestId, JSONObject jsonObject, Ticker ticker, CheckerInfo checkerInfo) throws Exception {
		String pairId = checkerInfo.getCurrencyPairId();
		if (pairId == null) {
			pairId = checkerInfo.getCurrencyBaseLowerCase()+"_"+checkerInfo.getCurrencyCounterLowerCase();
		}
		final JSONObject pairJsonObject = jsonObject.getJSONObject(pairId);
		ticker.bid = ParseUtils.getDoubleFromString(pairJsonObject, "buy");
		ticker.ask = ParseUtils.getDoubleFromString(pairJsonObject, "sell");
		ticker.vol = ParseUtils.getDoubleFromString(pairJsonObject, "volume");
		ticker.last = ParseUtils.getDoubleFromString(pairJsonObject, "rate");
	}

	// ====================
	// Get currency pairs
	// ====================
	@Override
	public String getCurrencyPairsUrl(int requestId) {
		return URL;
	}

	@Override
	protected void parseCurrencyPairsFromJsonObject(int requestId, JSONObject jsonObject, List<CurrencyPairInfo> pairs) throws Exception {
		final JSONArray pairIds = jsonObject.names();
		for(int i=0; i<pairIds.length(); ++i) {
			String pairId = pairIds.getString(i);
			if(pairId==null)
				continue;
			String[] currencies = pairId.split("_");
			if(currencies.length != 2)
				continue;

			pairs.add(new CurrencyPairInfo(
					currencies[0].toUpperCase(Locale.US),
					currencies[1].toUpperCase(Locale.US),
					pairId));
		}
	}
}

File: dataModule/src/main/java/com/aneonex/bitcoinchecker/datamodule/model/market/Bitso.kt
package com.aneonex.bitcoinchecker.datamodule.model.market

import com.aneonex.bitcoinchecker.datamodule.model.CheckerInfo
import com.aneonex.bitcoinchecker.datamodule.model.CurrencyPairInfo
import com.aneonex.bitcoinchecker.datamodule.model.Market
import com.aneonex.bitcoinchecker.datamodule.model.Ticker
import com.aneonex.bitcoinchecker.datamodule.util.TimeUtils
import org.json.JSONObject
import java.util.*

class Bitso : Market(NAME, TTS_NAME, null) {
    companion object {
        private const val NAME = "Bitso"
        private const val TTS_NAME = NAME
        private const val URL = "https://api.bitso.com/v3/ticker?book=%1\$s"
        private const val URL_CURRENCY_PAIRS = "https://api.bitso.com/v3/available_books/"
    }

    override fun getCurrencyPairsUrl(requestId: Int): String {
        return URL_CURRENCY_PAIRS
    }

    @Throws(Exception::class)
    override fun parseCurrencyPairsFromJsonObject(requestId: Int, jsonObject: JSONObject, pairs: MutableList<CurrencyPairInfo>) {
        val books = jsonObject.getJSONArray("payload")
        for (i in 0 until books.length()) {
            val pairId = books.getJSONObject(i).getString("book")
            val currencies = pairId.split("_".toRegex()).toTypedArray()
            if (currencies.size != 2) continue
            pairs.add(
                CurrencyPairInfo(
                    currencies[0].uppercase(Locale.ROOT),
                    currencies[1].uppercase(Locale.ROOT),
                    pairId
                ))
        }
    }

    override fun getUrl(requestId: Int, checkerInfo: CheckerInfo): String {
        // Compatibility with previous version
        val pairId: String = checkerInfo.currencyPairId ?: "${checkerInfo.currencyBaseLowerCase}_${checkerInfo.currencyCounterLowerCase}"
        return String.format(URL, pairId)
    }

    @Throws(Exception::class)
    override fun parseTickerFromJsonObject(requestId: Int, jsonObject: JSONObject, ticker: Ticker, checkerInfo: CheckerInfo) {
            jsonObject.getJSONObject("payload").let {
            ticker.high = it.getDouble("high")
            ticker.low = it.getDouble("low")

            ticker.bid = it.getDouble("bid")
            ticker.ask = it.getDouble("ask")

            ticker.vol = it.getDouble("volume")
            ticker.last = it.getDouble("last")

            ticker.timestamp = TimeUtils.convertISODateToTimestamp(it.getString("created_at"))
        }
    }

    @Throws(Exception::class)
    override fun parseErrorFromJsonObject(requestId: Int, jsonObject: JSONObject, checkerInfo: CheckerInfo?): String? {
        return jsonObject.getJSONObject("error").getString("message")
    }
}